File: fabric/src/main/java/net/chariskar/breakthemod/fabric/client/commands/lastSeen.java
/*
 * This file is part of breakthemod.
 *
 * breakthemod is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * breakthemod is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with breakthemod. If not, see <https://www.gnu.org/licenses/>.
 */

package net.chariskar.breakthemod.fabric.client.commands;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import com.mojang.brigadier.context.CommandContext;
import net.chariskar.breakthemod.Services.lastSeenService;
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource;
import net.minecraft.text.Text;


public class lastSeen extends FabricCommand {

    lastSeenService Service = new lastSeenService();

    @Override
    public String getName() {
        return "lastseen";
    }

    @Override
    public String getDescription() {
        return "Tells you when a person was last online";
    }

    @Override
    public String getUsageSuffix() {
        return "<name>";
    }

    @Override
    protected int execute(CommandContext<FabricClientCommandSource> ctx) throws Exception {
        String name = ctx.getArgument("name", String.class);

        // NOTE: client.exec is used here as this isnt a resource heavy task.
        client.execute(()->{
            String resp = Service.get(name);
            sendMessage(client, Text.literal(resp));
        });
        return 0;
    }


    @Override
    public void register(CommandDispatcher<FabricClientCommandSource> dispatcher) {
        dispatcher.register(
                LiteralArgumentBuilder.<FabricClientCommandSource>literal(getName())
                        .then(RequiredArgumentBuilder.<FabricClientCommandSource, String>argument("name", StringArgumentType.string())
                                .executes(context -> {
                                    if (!getEnabled()) return 0;
                                    return run(context);
                                })
                        )
        );
    }
}

File: src/main/kotlin/net/chariskar/breakthemod/client/commands/lastSeen.kt
package net.chariskar.breakthemod.client.commands

import com.mojang.brigadier.Command as command
import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.builder.RequiredArgumentBuilder
import com.mojang.brigadier.context.CommandContext
import kotlinx.coroutines.launch
import net.chariskar.breakthemod.client.api.Command
import net.chariskar.breakthemod.client.api.types.Resident
import net.chariskar.breakthemod.client.utils.ServerUtils.getEnabled
import net.chariskar.breakthemod.client.utils.Timestamps
import net.fabricmc.fabric.api.client.command.v2.FabricClientCommandSource
import net.minecraft.text.Text
import net.minecraft.util.Formatting

class lastSeen : Command() {
    init {
        name = "lastSeen"
        description = "Shows the last time a user was online"
        usageSuffix = "<name>"
    }

    override fun execute(ctx: CommandContext<FabricClientCommandSource>): Int {
        val name: String = ctx.getArgument("name", String::class.java)
        scope.launch {
            val player: Resident? = fetch.getResident(name)
            if (player == null) {
                sendMessage(Text.literal("Unable to find $name"), Formatting.RED)
                return@launch
            }
            val timestamps: List<Long> = Timestamps.parseTimestamp(player.timestamps?.lastOnline!!)
            if (player.status?.isOnline == true) {
                sendMessage(Text.literal("${player.name} has been online right now, for ${timestamps[0]} days, ${timestamps[1]} hours and ${timestamps[2]} minutes."), Formatting.AQUA)
            } else {
                sendMessage(Text.literal("${player.name} was last online ${timestamps[0]} days, ${timestamps[1]} hours and ${timestamps[2]} minutes."), Formatting.AQUA)
            }
            return@launch
        }
        return 0
    }

    override fun register(dispatcher: CommandDispatcher<FabricClientCommandSource>) {
        dispatcher.register(
            LiteralArgumentBuilder.literal<FabricClientCommandSource>(name)
                .then(
                    RequiredArgumentBuilder.argument<FabricClientCommandSource?, String>("name", StringArgumentType.string())
                        .executes(command { context: CommandContext<FabricClientCommandSource> ->
                            if (!getEnabled()) return@command 0
                            return@command run(context)
                        })
                )
        )
    }
}